### PR TITLE
Add float counter overloads

### PR DESCRIPTION
**Pull request by TristanRhodes**

### CODE REVIEW

File: StatsdClient/Statsd.cs [hunk 1]
```diff
@@ -181,6 +201,25 @@ private void InitialiseInternal(Func<IOutputChannel> createOutputChannel, string
         {
             if (value < 0)
             {
+                throw new ArgumentOutOfRangeException();
```
Comment (niemyjski) [inline StatsdClient/Statsd.cs]: I don't think this should throw an exception, only log it and return like the others. Stats client really shouldn't be blowing up.

File: StatsdClient/StatsdClientExtensions.cs [hunk 2]
```diff
@@ -14,7 +14,14 @@ public static class StatsdClientExtensions
         /// <param name="count">The counter value (defaults to 1).</param>
         public static void LogCount(this IStatsd client, string name, int count = 1)
         {
-            client.LogCountAsync(name, count).Wait();
+            try
```
Comment (niemyjski) [inline StatsdClient/StatsdClientExtensions.cs]: Can you revert this change, the aggregate exception should be boiled up as it can contain multiple exceptions.

File: StatsdClient/StatsdClientExtensions.cs [hunk 3]
```diff
@@ -24,7 +31,14 @@ public static void LogCount(this IStatsd client, string name, int count = 1)
         /// <param name="milliseconds">The duration, in milliseconds, for this metric.</param>
         public static void LogTiming(this IStatsd client, string name, long milliseconds)
         {
-            client.LogTimingAsync(name, milliseconds).Wait();
+            try
+            {
```
Comment (niemyjski) [inline StatsdClient/StatsdClientExtensions.cs]: Can you revert this change, the aggregate exception should be boiled up as it can contain multiple exceptions.

File: StatsdClient/StatsdClientExtensions.cs [hunk 4]
```diff
@@ -34,7 +48,48 @@ public static void LogTiming(this IStatsd client, string name, long milliseconds
         /// <param name="value">The value for this gauge</param>
         public static void LogGauge(this IStatsd client, string name, int value)
         {
-            client.LogGaugeAsync(name, value).Wait();
+            try
+            {
```
Comment (niemyjski) [inline StatsdClient/StatsdClientExtensions.cs]: Can you revert this change, the aggregate exception should be boiled up as it can contain multiple exceptions.

File: StatsdClient/StatsdClientExtensions.cs [hunk 5]
```diff
@@ -34,7 +48,48 @@ public static void LogTiming(this IStatsd client, string name, long milliseconds
...
+        /// <summary>
+        /// Log a gauge.
+        /// </summary>
+        /// <param name="name">The metric name</param>
+        /// <param name="value">The value for this gauge</param>
+        public static void LogGauge(this IStatsd client, string name, double value)
+        {
+            try
+            {
+                client.LogGaugeAsync(name, value).Wait();
```
Comment (niemyjski) [inline StatsdClient/StatsdClientExtensions.cs]: Can you revert this change, the aggregate exception should be boiled up as it can contain multiple exceptions.

File: StatsdClient/StatsdClientExtensions.cs [hunk 6]
```diff
@@ -34,7 +48,48 @@ public static void LogTiming(this IStatsd client, string name, long milliseconds
...
+        /// Log a gauge.
+        /// </summary>
+        /// <param name="name">The metric name</param>
+        /// <param name="value">The value for this gauge</param>
+        public static void LogGauge(this IStatsd client, string name, decimal value)
+        {
+            try
+            {
+                client.LogGaugeAsync(name, value).Wait();
+            }
```
Comment (niemyjski) [inline StatsdClient/StatsdClientExtensions.cs]: Can you revert this change, the aggregate exception should be boiled up as it can contain multiple exceptions.

File: StatsdClient/StatsdExtensions.cs [hunk 7]
```diff
@@ -49,6 +49,55 @@ private class StatsBuilderInternal : DynamicObject
...
+            case MetricType.TIMING:
+              client.LogTiming(name, value);
+              break;
+            case MetricType.SET:
+              client.LogSet(name, value);
+              break;
+          }
+      }
+
+      private static void DoubleHandler(IStatsd client, string metricType, string name, object arg)
```
Comment (niemyjski) [inline StatsdClient/StatsdExtensions.cs]: You're potentially losing precision by using a double as the default rather than a decimal.

**Comment by niemyjski** [comment]
Thanks for your PR, I've added some comments to it :)

**Comment by TristanRhodes** [comment]
Huh. I didn't mean to PR the fork, was doing the PR to master for a colleague. I think I just failed at GitHub :)

I spent a while making the code work with the tests, I'm happy to roll back the changes, but I'll also need to update the tests. How do you want that done?
Cheers
Tristan

**Comment by niemyjski** [comment]
I think it all looked good, just don't think that we should be throwing exceptions for stats if the number is less than 0.

**Comment by niemyjski** [comment]
I don't think we should be rolling back any tests :) unless they are checking for an exception to be thrown.. @lukevenediger @TristanRhodes what do you think?

**Comment by TristanRhodes** [comment]
I made the changes to get the tests here passing:
https://github.com/lukevenediger/statsd-csharp-client/blob/master/StatsdClientTests/StatsdTests.cs#L36

They are all expecting an ArgumentOutOfRangeException or an ArgumentNullException, but I think due to the refactoring to make the statsD client Async, they are now all catching AggregateExceptions which is causing the tests to fail. (At least it was in the NCrunch runner).

I can revert and let it throw the AggregateExceptions but then we will need to update the tests to drill into that and check that it has the correct internal exception.

Some tests also were expecting exceptions which did not occur as the throw seems to have been replaced with Trace.Log.

Let me know if you're happy for me to change the tests to work with the original code, excluding the float / double change.

**Comment by niemyjski** [comment]
The solution is to do `.GetAwaiter.GetResult();` instead of `.Wait();` && `.Result`. This will cause it to block but bubble up the real exception.

Also, anytime we are doing a `GetResult` we should be calling `.ConfigureAwait(false).GetAwaiter.GetResult()` in the core library (never do ConfigureAwait(false) in UI or test projects). If you update this, it should fix the issues you ran into 👍

**Comment by niemyjski** [comment]
@TristanRhodes any updates on this?

**Comment by niemyjski** [comment]
@TristanRhodes I just merged this in manually and gave you credit...
